### ISSUE TITLE
API proxy connection to backend cannot be secured

### ISSUE THREAD

**Issue by rhuss** [openshift/origin#4497]
We have the use case that we have monitoring agents installed on our Java pods ([Jolokia](http://www.jolokia.org) )which are accessed via the OpenShift API Proxy. This agent is used by the [fabric8](http://fabric8.io) console for a deep inspection of the managed pods (e.g. you can dive into the pod for debugging, visualisation of Camel routes, JMX operations, ...).

This agent can be secured via OAuth in that it can validate a given token against the OpenShift OAuth server. 

However, the OS API Server seems to [filter out](https://github.com/openshift/origin/blob/945fb793cf68ba3836a75239b14b05fcbe44427a/pkg/util/httpproxy/upgradeawareproxy.go#L237-L241) the `Authorization:` header before accessing the backend. I guess this is for security concerns to prevent a malicious agent to get access to the admin token. 

My question is whether there is an alternative solution for this use case ? It would possibly help to copy over another, custom header (like `X-Authorization:`) over to `Authorization:` before contacting the backend. On the other hand, if the client and the backend agree, the client can always push the token to the other side (eg. by using any non-filtered header or even via the payload). So maybe filtering the `Authorization:` header doesn't make so much sense ? Ok, this filter prevents unconditionally passing of the header to the backend when the client doesn't want to.

tl;dr - Is there a canonical way how a proxied service can participate in OAuth SSO ?

**Comment by liggitt** [comment]
Authentication to the OS API is distinct from accessing the applications... it is very intentional not to leak API credentials to applications running on the infrastructure.

**Comment by rhuss** [comment]
@liggitt completely agreed, but I wonder how the application behind the proxy could then be authenticated when the `Authorization:` slot is already occupied for securing the API access.

**Comment by liggitt** [comment]
if the API was ONLY acting as a proxy, I would expect to use `Proxy-Authorization` for the API, and `Authorization` for the back-end app, but since the same auth handler services the API and proxy endpoints, and the same client code passes authentication tokens to both, it's going to be tough to special case auth to the proxy endpoints

**Comment by liggitt** [comment]
if the app already has its own authorization, can't you just access it directly using a route->service that exposes it? I thought the reason we were going through the API was to get the benefit of the API's authorization

**Comment by rhuss** [comment]
Its also about the principal used for authorization in addition to authentication. This information gets also lost. For authentication alone IMO it would be sufficient to let proxy do the authentication alone. Maybe we could copy over the principal information into an header when proxying ?

**Comment by liggitt** [comment]
that's interesting... a REMOTE_USER header of sorts... the proxy could strip that header from requests and replace it with the authenticated principal, but I'm not sure who else has network visibility to the pods who could spoof that header

**Comment by rhuss** [comment]
> if the app already has its own authorization, can't you just access it directly using a route->service that exposes it?

We could do, but since its about an agent installed for managing/monitoring every (java based) service an own route feels a bit heavy weight. Also since it's something which we don't want to expose to everyone, too, the proxy based access for this monitoring/management feature feels just right.

**Comment by jimmidyson** [comment]
The main use case right now is the jmx console in the openshift console.
This connects to pods via the api server proxy & at the moment the jolokia
endpoint in each pods is unauthenticated. We want to enable auth &
preferably reuse openshift auth token.

**Comment by rhuss** [comment]
yes, something like a `REMOTE_USER` would be perfect (e.g. the OAuth user). /wrt the network visibility this could be indeed an issue if not going over SSL. The only 'strong' solution beyond using SSL I see would be signing a la JSON web tokens but thats probably over the top ....

**Comment by rhuss** [comment]
> The main use case right now is the jmx console in the openshift console.
> This connects to pods via the api server proxy & at the moment the jolokia
> endpoint in each pods is unauthenticated. We want to enable auth &
> preferably reuse openshift auth token.

@jimmidyson but when we allow only access via the proxy, couldn't we *trust* the proxy that he did the authentication ?

**Comment by liggitt** [comment]
SSL only helps you if the proxy uses a client cert to make the request and the app verifies it

**Comment by rhuss** [comment]
> SSL only helps you if the proxy uses a client cert to make the request and the app verifies it

That's true and of course also for the *authentication* trust relationship between proxy and proxied service.

**Comment by jimmidyson** [comment]
We could but that would mean changing the trusted proxy for every
installation, updating it masters are added, etc.

On 18:41, Thu, 3 Sep 2015 Roland Huß <notifications@github.com> wrote:

The main use case right now is the jmx console in the openshift console.
This connects to pods via the api server proxy & at the moment the jolokia
endpoint in each pods is unauthenticated. We want to enable auth &
preferably reuse openshift auth token.


@jimmidyson <https://github.com/jimmidyson> but when we allow only access
via the proxy, couldn't we *trust* the proxy that he did the authentication
?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/4497#issuecomment-137523745>.<img
src="
https://ci5.googleusercontent.com/proxy/PIyaQGcsyFi7vCpSpRFm9LAh2iEjbAijaYhHrbUUJTOxTxT0myuZT9wBoef_NPqSyPH39SwW41YdTcnHeZ4Tmdyj42y_z6OydBoMvXyjvn7CODE0n5IgwOET4lKdMHWJCpuFUfJ1en34Ho2j9LCICHWuggZPPA=s0-d-e1-ft#https://github.com/notifications/beacon/AAcXE3Z6-OSia2cG5Ge5I9Z1Fwvec9uFks5ouH3igaJpZM4F3Iby.gif
">

**Comment by jimmidyson** [comment]
A jwt would do the same job as a client cert wouldn't it? Still needs validation of course. But we've discussed this before. This is another use case I guess.

**Comment by liggitt** [comment]
it allows validation, but also hands credentials to the app, which I'm not happy about

**Comment by liggitt** [comment]
we'd want separate id/access tokens

**Comment by dhirajsb** [comment]
How about some kind of temporary token with user id, which is blacklisted for api access? Doesn't quite solve your problem of hiding id, but would block api attacks.

**Comment by rhuss** [comment]
@dhirajsb the token is provided and created by the OAuth server and doesn't contain user id information itself. You receive the user info from the auth server only as response to the token verification request. So its not so easy for the API server to create such temporary tokens. Also it would involve holding state in the API server (the black list) which I don't know if this is feasible.

**Comment by liggitt** [comment]
the proxy code is all upstream, which has no dynamic token system for normal users at the moment

**Comment by rhuss** [comment]
> it allows validation, but also hands credentials to the app, which I'm not happy about

You can use jwt with asymmetric, public key validation, too. The public key could be injected into the pods e.g. via envs. That way only the API server can validly sign the token (e.g. holding the username) and the backend service can verify it without sharing any credentials.

**Comment by jimmidyson** [comment]
OK so the token would be an ID token, rather than an access token? SGTM!

**Comment by rhuss** [comment]
Our current approach (probably more a workaround) to allow the transport of the authorization token between two trusted applications (i.e. the openshift / fabric8 console and Jolokia for deep inspection of a running Java container) by using a custom header (e.g. `X-Authorization`) doesn't work because this header can't be added to the CORS preflight check's response of the proxy itself. More details are described [here](https://github.com/hawtio/hawtio-jmx/issues/10).

Currently we are stucked for how to secure Jolokia as a backend, administrative service behind the API proxy. IMO the alternative to leave Jolokia unsecured is no alternative and much worse than passing the API token (since any other pod can reach these open ports). But I agree, allowing the admin token to be passed to any backend service is no solution either.

So, my question really boils down whether the use case

> **As** an administrator 
> I **want to** access a backend administrative service via the OpenShift proxy in a secured fashion
> **so** that I can perform application specific administrative tasks

is supported somehow ? (as mentioned above, exposing it as separate services for every pod is not really an option).

**Comment by smarterclayton** [comment]
@liggitt let's try to figure out what the minimal enablement we can do here is.

**Comment by liggitt** [comment]
There are several issues with plumbing authorization headers through the proxy to backend apps:
1. Conflicts with API authorization headers (as noted above)
* Headers are [automatically removed](https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/auth/authenticator/request/bearertoken/bearertoken.go#L35) after auth to avoid leaking credentials
2. Conflicts with OpenShift<->Kube authorization headers when running OpenShift against a separate Kube instance (client -> OpenShift -> Kube -> backend)
* Headers are [removed](https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/util/httpproxy/upgradeawareproxy.go#L243-L245) and potentially [replaced](https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/util/httpproxy/upgradeawareproxy.go#L264-L270) with the OpenShift credentials to proxy to Kube
3. API proxy support for https to the backend is not present (see https://github.com/kubernetes/kubernetes/issues/14194)

If the backend app wants to handle authorization, is there a reason not to expose it as a service and take the proxy out of the equation?

**Comment by smarterclayton** [comment]
Although to be fair proxy is there precisely to let you deal with the 20% -
everything else is busted.  Having the auth be able to come through a proxy
(not necessarily Kube auth, but a separate X-Authorization header) would at
least allow you to do that.

I agree though that pass through of the auth header is very scary.

On Sep 22, 2015, at 3:33 PM, Jordan Liggitt <notifications@github.com>
wrote:

There are several issues with plumbing authorization headers through the
proxy to backend apps:
1. Conflicts with API authorization headers (as noted above)

   - Headers are automatically removed
   <https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/auth/authenticator/request/bearertoken/bearertoken.go#L35>
   after auth to avoid leaking credentials
   - Conflicts with OpenShift<->Kube authorization headers when running
   OpenShift against a separate Kube instance (client -> OpenShift -> Kube ->
   backend)
   - Headers are removed
   <https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/util/httpproxy/upgradeawareproxy.go#L243-L245>
   and potentially replaced
   <https://github.com/openshift/origin/blob/dc1a7b07d54c2c169b9f278cb5ee05e86d5011c6/pkg/util/httpproxy/upgradeawareproxy.go#L264-L270>
   with the OpenShift credentials to proxy to Kube
   - API proxy support for https to the backend is not present (see
   kubernetes/kubernetes#14194
   <https://github.com/kubernetes/kubernetes/issues/14194>)

If the backend app wants to handle authorization, is there a reason not to
expose it as a service and take the proxy out of the equation?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/4497#issuecomment-142393782>.

**Comment by liggitt** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/14876 to address the issue with proxying TLS to the backend pods

**Comment by liggitt** [comment]
Created https://trello.com/c/kTX8fBNs/522-allow-api-proxy-to-securely-connect-to-backend-pods to track overall progress
